### DISCUSSION TITLE
How is bug fixing being prioritised?

### DISCUSSION THREAD

**Discussion by williambuchanan2** [dotnet/maui#21607] [General]
Just wondering what the logic is behind the prioritising of bug fixing...

From what I can see, a lot of critical bugs which have been repeatedly reported are getting ignored, while at the same time relatively minor bugs reported only 2 weeks ago are already being worked on in the current iteration.

I have these 2 critical issues causing major problems which actually prevent the app from working:

https://github.com/dotnet/maui/issues/19383 

https://github.com/dotnet/maui/issues/19175 

Both were reported in December - but had also been reported in various other forms before then (particularly the collection view scrolling issue which has a lot of people complaining about it). 

There is also the well known ios keyboard issue that everyone has been screaming about for at least 18 months - it's still not fixed.

Surely things like this would be more important to address than a lot of the things which seem to be in the current iteration?

**Comment by jonmdev** [comment]
My guess: There is no particular logic. There are too many bugs and too few people. Many or most of the very few people working on the bug fixes probably had no involvement in the original creation of Maui. Maui is complicated with many moving parts.

For example, the team have said they know that Windows apps randomly start to blank screens but they can't find the cause [here](https://github.com/dotnet/maui/issues/15533). That was reported 10 months ago now.

So the 5-10 people on the team doing bug fixes likely just pick random ones they think they can fix and work on those until they hit a wall or solve it, then move on to the next. 

I used to do a lot of Unity development. I used to think it was bad that they would take 6 months over there to fix a verified bug. But they would fix anything you proved was related to their system with a clear bug report and repro project. Extremely good.

Now I realize how spoiled I was. 😂

I still can't see how Microsoft can fix all the problems without twice or more the workforce assigned to it currently. There is only so much time in the day and it is hard to track down strange bugs in complex software.

**Reply by williambuchanan2** [reply]
I agree - there doesn't seem to be any particular logic from what I can see. Also, to me it looks like there is nobody managing it - if there was I would expect someone to be taking charge to make sure critical issues are prioritised, rather than wasting their time fixing stupid problems like corners on borders not being properly round, etc! 

Is there nobody in Microsoft responsible for this epic failure? I just don't get it....

**Comment by Vroomer** [comment]
It's been stated multiple times that one of the way the team determines importance of an issue is the count of likes. Your issues have 0 and 1 like respectively. Why would you expect such issues to be prioritized, when they are not affecting many users?

**Reply by MAUIoxo** [reply]
In my opinion it should be prioritized so that the issues that are reported on the basic controls (like CollectionView, SwipeItemView, …) and layouts get fixed soon so that we get the impression the system is stable. I had difficult times in the past with things I had to work around which cost me a lot of time. Issues that were things that used to work before and stopped working or things that work on iOS but not on Android or vice versa. And in my opinion the basic things that do not work in simple scenarios like first Flyout Item not working anymore when there was an Header in the Flyout which came as a regression, these are the things that must not happen and should be fixed asap in my opinion, so that there is a fix at least in a month. Just my 50 cents on that…

**Reply by jonmdev** [reply]
That is my perspective as well. Like count is not a reliable indicator of the significance of a bug. Many people may be affected by a bug without knowing they are affected by the bug.

Ie. If there is a bug affecting the performance of a simple views function then we are all affected whether different individuals can see it or not.

For example, many people complain about the scrolling performance of collectionview. However the problem is almost certainly not in collectionview itself (based on evidence and analysis from multiple users) but rather the underlying core components themselves. Ie. Collectionview can't scroll smoothly because the core components and/or layout system are updating improperly and inefficiently.

Thus an obscure complaint about the core controls may hold the still currently unknown solution to the collectionview problem. Many esoteric bugs that people cannot understand the source for may possibly be fixed by correcting any known issues with the core controls.

I think the core elements are: border, label, absolute layout, image, editor, and entry.

With that all said, if you look at the issues in progress I don't see anything that looks insignificant. It just seems there are still many bugs:

https://github.com/orgs/dotnet/projects/194

One must hope that some of the more esoteric or strange bugs will disappear as the most basic functions are fixed.

I had a few bugs that disappeared without any specific attention in just this manner through version updates.

**Reply by MAUIoxo** [reply]
Also not to forget Grid and ScrollView where I still have strange behaviors at certain points like in [#17291](https://github.com/dotnet/maui/issues/17291). Still got this last week and this was reported in September. In my opinion this is why so many people complain about it. Well, the argument could be that we could all help to fix bugs which is correct in a way. But, I think a certain expectation that the basic things should work right and don’t get easily destroyed like in the Flyout and SwipeViewItem examples is allowed. We don’t necessarily have the time to fix the stuff ourselves as we all try to realize our own projects while reporting misbehaviors. When I count the time just for reporting issues and working around it that’s also a lot from my spare time, too

**Reply by williambuchanan2** [reply]
@Vroomer 
https://github.com/dotnet/maui/issues/19461 - 0 likes - currently being worked on.
https://github.com/dotnet/maui/issues/21173 - 0 likes - reported 3 weeks ago. Currently being worked on.
https://github.com/dotnet/maui/issues/21314 - 0 likes - reported 3 weeks ago. Currently being worked on.
https://github.com/dotnet/maui/issues/21034 - 0 likes - reported 4 weeks ago. Currently being worked on.

I could go on...

However, for me the problem is that, even if it was based on likes, using likes is a terrible way to prioritise. Most devs that I know don't bother liking the bug reports. They see that someone else has reported the problem so they don't bother doing anything. I have tried to get my team to report problems but most of the time they can't be bothered to even raise bugs - especially now that they feel like it's a waste of time because nobody is fixing them anyway.

Also, on the subject of collection/scroll view et al - these problems have been reported multiple times in different forms by different people. If each of those reports only get 1 like each then the likelihood is that they will never be prioritised because the simple bugs which have 2 likes will jump to the top.

It just feels like the whole priority system is flawed and critical issues are not getting addressed as a result...

**Reply by MAUIoxo** [reply]
The question is who is decider how it is prioritized? @davidortinau

**Reply by davidortinau** [reply]
I've added a high level statement about prioritization of issues for the current servicing of .NET 8 to the roadmap https://github.com/dotnet/maui/wiki/Roadmap.

Xamarin migrations are the top priority with end of support coming May 1. Looking at those 5 issues @williambuchanan2 listed, they are all impacting projects in this category. The engineers assigned to them are either recently joining the core team from other teams and opportunistically working issues they are able to resolve, or they are working issues within the focus area they are best suited to resolve.

We also track issue engagement (comments, number of commenters, reactions, etc.) and factor than in when planning a sprint.

**Reply by MAUIoxo** [reply]
@davidortinau thanks for adding that up and making it clear. Also thanks for the additional information about people that can currently support

**Reply by williambuchanan2** [reply]
@davidortinau How many developers are actively working on fixing bugs? I see .NET 9 is now coming up - last year we all waited for bug fixes but this didn't happen because all the focus went on .NET 8 - which didn't fix any of the critical problems. If there aren't enough people working on bug fixing (which obviously is the situation) it wouldn't be hard to imagine that the same thing is going to happen again this year...
